README.md
## 简介

顾名思义，这是一个可以在终端执行的 hutool。

那 hutool-cli 可以做什么呢？

首先Hutool是一个非常好用的Java工具库，提供了非常多地静态工具方法供大家使用，极大地提高了我们日常的开发效率；
但是如果我们想要快速的知道一个方法的运行效果，或者是调试一些方法，亦或者是想利用一些工具方法生成一些内容并复制它到其他应用程序，那我们就不得不写个测试类，然后再调一下对应的方法，最后查看执行结果，但这未免显得过于费劲。

于是 hutool-cli 把 hutool 带到了终端中，让我们可以直接通过命令去执行一个静态的工具方法，让日常的开发变得更有效率。

比如我们想生成一个随机UUID，现在只需要打开终端执行下面命令：

```shell
hu uuid
# output: 483cc7fc-4b22-4188-8f1c-dc1ce4b6d3ee
```

随着工具不断的迭代更新，现在，已经不再只是终端里的Hutool了，更多功能等待你来探索哦。

## 内置命令

jq命令支持（jq是一款强大的命令行json数据处理工具）

```shell
hu jq '.' '{"tutorial":"https://stedolan.github.io/jq/tutorial/"}'

#output:
{
  "tutorial" : "https://stedolan.github.io/jq/tutorial/"
}

# 其他示例：
hu jq 'to_entries|map(.key+"="+.value.clazz)|.[]' hutool/class.json strip
```

> [JQ基本语法](https://stedolan.github.io/jq/manual/#Basicfilters)

文件查找

```shell
# 命令格式
hu find {dir} {nameFilter} [(file|dir)] [hidden] [ignoreempty] [depth:{n}] [(c|a|u)time+{n}{timeUnit}]

# 举个例子：
hu find . java file depth:9 utime+7d hidden # 查找当前目录下文件名包含java并且最近7天修改过的文件，遍历最大深度为9层目录，hidden表示允许查找隐藏文件
hu find ~ '.*' dir ctime:9d ignoreempty # 查找用户目录下最近9天创建的，且不为空的文件夹
```

> 说明：{}表示必传变量，[]表示可选参数，()表示枚举值，可选参数不区分位置，时间单位参考最后面的日期参数格式。

FigletBanner生成（ASCII艺术字）

```shell
hu figlet hutool-cli
# hu figlet {message} {font-name=standard}

# output:
  _               _                     _                  _   _ 
 | |__    _   _  | |_    ___     ___   | |           ___  | | (_)
 | '_ \  | | | | | __|  / _ \   / _ \  | |  _____   / __| | | | |
 | | | | | |_| | | |_  | (_) | | (_) | | | |_____| | (__  | | | |
 |_| |_|  \__,_|  \__|  \___/   \___/  |_|          \___| |_| |_|
                                                                 

```

> [figlet fonts](http://www.figlet.org/fontdb.cgi) ，下载后放到 `{user.home}/hutool-cli/fonts/figlet/` 目录下即可。

查看时间进度

```shell
hu dayp

# output:
辛丑牛年 五月初三 周六 2021-06-12 09:45:07.755

今日 [oooooooooooooooooooooooooooooooooooooooo                                                            ]: 40.63%
本周 [oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo                        ]: 76.79%
本月 [oooooooooooooooooooooooooooooooooooooooo                                                            ]: 40.00%
本年 [oooooooooooooooooooooooooooooooooooooooooooo                                                        ]: 44.66%
```

列出目录文件

```shell
hu ls

# output:
2021-05-30 00:46:38	      0	bin
2021-06-02 19:48:08	      0	external
2021-06-12 10:29:19	   4 KB	method
2021-06-25 19:26:38	4.21 KB	class.json
2021-06-25 19:27:11	8.06 KB	command.json
2021-06-12 10:29:19	  749 B	converter.json
2021-06-09 22:19:07	  165 B	external.conf
2021-06-24 22:15:32	 7.2 MB	hutool.jar
```

查看日历

```shell
hu calendar 202106

# output:
     2021-06-25    
Mo Tu We Th Fr Sa Su
    1  2  3  4  5  6
 7  8  9 10 11 12 13
14 15 16 17 18 19 20
21 22 23 24 25 26 27
28 29 30               

# also support like follow command: 
# hu calendar 6
# hu calendar 2020
# hu calendar 202101,2,3,4,5
# hu calendar 6,7,8
```

文件树

```shell
hu tree

# output:
├─bin
│ └─hu.exe
├─external
├─method
├─class.json
├─command.json
├─converter.json
├─external.conf
└─hutool.jar
```

计算倒计时

```shell
hu countdown 123456789 ms
# 时间单位可以参考最下方的对照表

# output:
1天10小时17分36秒789毫秒
```

## ZIP包安装

所需环境

- git
- java 8+

下载本项目

```shell
git clone https://gitee.com/code4everything/hutool-cli.git
```

下载对应的 [ZIP包](http://share.qiniu.easepan.xyz/tool/hutool/hu-1.5.zip) ，目录结构如下

```text
├─bin
│ ├─hu
│ ├─hu-mac
│ └─hu.exe
└─hutool.jar
```

下载完成后解压ZIP包，并将 hutool.jar 和 bin目录中与平台对应的可执行文件移动到 hutool-cli/hutool 目录下，如windows对应的`hu.exe`文件，linux对应的`hu`
文件，macos对应的`hu-mac`文件（Mac移动后需重命名为`hu`），移动后目录结构如下。

## 源码安装

> 也可用于更新 `hutool.jar` 文件，当已有 `bin` 目录文件，更新版本时无需再下载上面的 `zip` 包。

先拉取最新代码：`git pull`，然后执行打包命令，通常情况我们只需要执行PACK打包命令即可。

- `gradle pack -x test` 打包可执行JAR包，需要JAVA环境。
- `gradle install -x test` 执行上面的PACK打包命令，并构建对应平台的二进制执行文件（即`hu(.exe)`），需要GO语言环境。

> `-x test` 表示跳过测试用例。

## 配置环境变量

新建 `HUTOOL_PATH` 变量，对应的路径如下：`your_path/hutool-cli/hutool`，目录结构如下：

```text
├─bin
│ └─hu(.exe)
├─method
├─class.json
├─command.json
└─hutool.jar
```

最后将 `HUTOOL_PATH`/bin 添加到系统变量 `PATH` 中。

## 如何使用

查看支持的指令

```shell
hu

# output:
Usage: hutool-cli [options]
  Options:
    -r, --run, --command
      命令模式，命令（方法别名）可以精确的定位到静态方法，-r指令可缺省
    -c, --class
      类名称（自动加前缀：cn.hutool.）
    -m, --method
      方法名
    -t, --type
      参数类型（非必须，缺失时可能导致无法精确的定位方法）
    -p, --param, --parameter
      方法需要的参数
    -y, --yank, --copy
      是否将结果复制到剪贴板
    -a, --auto-param
      将剪贴板字符串内容注入指定索引位置的参数
    -v, --version
      查看当前版本
    -d, --debug
      是否开启调试模式
```

查看版本

```shell
hu -v
# output: hutool-cli: v1.2
```

执行一个方法

```shell
hu -c cn.hutool.core.util.IdUtil -m randomUUID
# output: 3214683f-55c1-412e-8b7a-454c57468d99

hu -r cn.hutool.core.codec.Base64#encode -t java.lang.CharSequence -p hutool-cli
# output: aHV0b29sLWNsaQ==
```

通过别名执行

```shell
hu -c base64 -m encode -p 'sky is blue'
# output: c2t5IGlzIGJsdWU=

hu encode64 'sky is blue' -y
# output: c2t5IGlzIGJsdWU=
# 说明：-y 表示将输出结果复制到剪贴板

hu decode64 -a:0
# output: sky is blue
# 说明：-a:0 表示将剪贴板字符串内容注入到索引位置是0的参数中

# v1.2新功能：支持连续执行，符号 // 分隔多个命令，\\0表示将第一个执行结果作为参数注入，\\1表示将第二个执行结果作为参数注入，依次类推
hu encode64 'test multi cmd' // decode64 \\0
# output: test multi cmd
```

> 在 `-r` 模式下，别名后可直接跟方法需要的参数，当然使用 `-p` 也是支持的，并且 -r 是可以省略的，如最上面生成随机UUID的例子。

## 查看类有哪些可执行静态方法

```shell
hu methods regex

# regex 是类 cn.hutool.core.util.ReUtil 的别名
# output:
contains(pattern:java.util.regex.Pattern, content:java.lang.CharSequence)
contains(regex:java.lang.String, content:java.lang.CharSequence)
count(pattern:java.util.regex.Pattern, content:java.lang.CharSequence)
count(regex:java.lang.String, content:java.lang.CharSequence)
..........
```

> v1.2版本已支持输出方法形参名称

## 别名查看

查看有哪些方法别名（可以精确的定位到方法）

```shell
# 查看包含random的别名
hu alias random

# output:
random     = cn.hutool.core.util.RandomUtil#randomString(length:int=16)
randomc    = cn.hutool.core.img.ImgUtil#randomColor()
randomi    = cn.hutool.core.util.RandomUtil#randomLong(min:long=0, max:long=9223372036854775807)
randompass = cn.hutool.core.util.RandomUtil#randomString(number:java.lang.String=abcdefghijkmnpqrstuvwxyzABCDEFGHJKMNPQRSTUVWXYZ234567892345678923456789, i:int=16)
uuid       = cn.hutool.core.util.IdUtil#randomUUID()
```

查看类名称别名

```shell
# 查看包含base的类别名
hu -c alias -p base

# output:
base32 = cn.hutool.core.codec.Base32
base64 = cn.hutool.core.codec.Base64
```

> 我们可以通过关键字 `alias` 来查看命令、类名、方法名已有的别名

## 覆盖别名方法

使用别名文件定义的类名和方法名，参数类型及数量替换为用户在终端中输入的内容。

举个例子，比如别名文件中定义的`md5`方法是：`cn.hutool.crypto.SecureUtil#md5(data:java.lang.String)`，传入的是一个`string`类型的参数，但是现在我们想要计算一个文件的MD5。

如何做呢？方法一，再定义一个计算文件MD5的别名，缺点导致别名过多，不方便记忆。

方法二，采用覆盖别名方法，语法：`hu command@type1,type2 params`。

还是上面的`md5`举例，首先查看它有哪些重载方法

```shell
hu md5@

# output:
md5()
md5(data:java.io.InputStream)
md5(data:java.lang.String)
md5(dataFile:java.io.File)
```

我们可以看到该方法是支持传入文件的，这时我们直接使用覆盖别名即可：

```shell
# file 是 java.io.File 的别名
hu md5@file command.json

# output:
268f6d715cf4ec191a96b80c29f2449f
```

> `hu command@` 查看别名的重载方法。

## 默认值

我们可以在别名文件中定义参数默认值，执行方法时默认值要么都缺省，要么都填上。

举个例子，如生成二维码图片：

查看生成二维码的命令

```shell
# 查找qrcode相关命令
hu alias qr

# output:
decodeqr = cn.hutool.extra.qrcode.QrCodeUtil#decode(qrCodeFile:java.io.File)
qrcode   = cn.hutool.extra.qrcode.QrCodeUtil#generate(content:java.lang.String, width:int=1000, height:int=1000, targetFile:java.io.File=qrcode.png)
```

然后执行

```shell
# 该方法有默认值，默认值可全缺省
hu qrcode 'qrcode test' /home/test.png
```

或者不使用默认值

```shell
# 虽然该方法有默认值，但我们可以不使用
hu qrcode 'qrcode test' 600 600 /home/test.png
```

## 自定义别名

别名可以用来快速的指向一个类或一个静态方法，格式大致如下，json中的key将作为别名，value包括了别名对应的类名、方法名以及方法所需的参数类型。

```json
{
    "encode64url": {
        "method": "cn.hutool.core.codec.Base64#encodeUrlSafe",
        "paramTypes": [
            "j.char.seq"
        ],
        "outArgs": "--copy"
    },
    "echo": {
        /*@符号表示总是解析参数默认值*/
        "method": "qe#run(@string=return args,@boolean=false)",
        "outConverter": "org.code4everything.hutool.converter.LineSepConverter",
        "inConverters": []
    }
}
```

> v1.2支持简写部分Java常用类名，滚动至底部查看更多。

hutool-cli 提供了很多常用的别名，参考下面文件：

- 方法别名参考 [command.json](/hutool/command.json)

- 类名称别名参考 [class.json](/hutool/class.json)

自定义你的别名后，可以 pr 到本仓库哦，让更多人体验到工具带来的便捷~

> 注意：类别名的定义不能超过16个字符。

> v1.2支持定义私有别名啦（优先级高于系统别名），定义路径 `{user.home}/hutool-cli/`，文件名和别名格式参照上面说明，程序会对私有别名和系统别名进行合并。

> v1.6新增在别名文件中定义参数输入输出转换器 `outConverter` `inConverters`，以及输出参数 `outArgs`。

## 加载外部类

方法一，在 `HUTOOL_PATH` 对应的目录下新建external目录，将类class类文件（包含包名目录）拷贝到external文件夹中即可，如类 `com.example.Test`
对应的路径 `external/com/example/Test.class`。

方法二，在 `HUTOOL_PATH` 对应的目录下新建external.conf文件，在文件中定义类加载路径（不包含包路径），使用英文逗号或换行符分隔，还是用上面的类举例，假设类绝对路径是 `/home/java/com/examaple/Test.class`，那么文件定义路径 `/home/java` 即可。

external.conf文件支持mvn坐标，但前提是本地maven仓库已有对应的jar包，比如：

```txt
// 这是注释
/path/folder
/path/test.jar
mvn:org.code4everything:wetool-plugin-support:1.6.0
```


CHANGE_LOG.md
## 变更记录

### v1.6

- 支持在别名文件中定义输入输出转换器
- 支持在别名文件中定义输出参数
- 优化别名读取方式
- 新增插件支持
- 移除不常用别名
- 支持生成FigletBanner
- 支持tree zip文件
- 新增Color类转换器
- 优化别名过滤
- 新增jq命令支持
- 新增帮助信息
- 删除了一些方法
- 新增unix常用命令支持
- 支持定义默认指令

### v1.5

- 优化方法调用
- 支持external.conf添加备注
- 新增string2json方法
- 修复在Java8平台无法执行的问题

### v1.3

- 支持参数数组
- 新增日期日历查看
- 新增文件树查看
- 优化默认值
- 优化日期转换器
- 支持日期运算
- 新增倒计时命令
- 支持别名搜索
- 支持执行QL脚本
- 新增生成双色球号码
- 支持覆盖别名方法
- **切换到Kotlin**

### v1.2

- 修复了一些已知问题
- 新增常用别名
- 支持私有别名
- 支持输出静态方法形参名称
- 新增Java常用类名简写
- 支持多命令执行
- 支持加载外部类

### v1.1

- 优化方法根据参数个数自动匹配
- 优化剪贴板内容作为参数注入
- 支持参数默认值

> 下载 [windows](http://share.qiniu.easepan.xyz/tool/hutool/windows-1.1.zip) [linux](http://share.qiniu.easepan.xyz/tool/hutool/linux-1.1.zip) [macos](http://share.qiniu.easepan.xyz/tool/hutool/darwin-1.1.zip)

### v1.0

- 支持格式化输出
- 支持复制执行结果
- 支持剪贴板内容作为参数
- 支持别名
- 通过反射执行静态方法

> 下载 [windows](http://share.qiniu.easepan.xyz/tool/hutool/windows-1.0.zip) [linux](http://share.qiniu.easepan.xyz/tool/hutool/linux-1.0.zip) [macos](http://share.qiniu.easepan.xyz/tool/hutool/darwin-1.0.zip)


src/main/java/org/code4everything/hutool/Utils.java
package org.code4everything.hutool;

import cn.hutool.core.comparator.ComparatorChain;
import cn.hutool.core.date.ChineseDate;
import cn.hutool.core.date.DateTime;
import cn.hutool.core.date.DateUtil;
import cn.hutool.core.date.Week;
import cn.hutool.core.exceptions.ExceptionUtil;
import cn.hutool.core.io.FileUtil;
import cn.hutool.core.lang.Holder;
import cn.hutool.core.lang.JarClassLoader;
import cn.hutool.core.math.Calculator;
import cn.hutool.core.util.ReflectUtil;
import cn.hutool.core.util.StrUtil;
import com.alibaba.fastjson.JSONObject;
import javassist.ClassPool;
import javassist.CtClass;
import javassist.CtMethod;
import javassist.NotFoundException;
import javassist.bytecode.LocalVariableAttribute;
import org.code4everything.hutool.converter.ClassConverter;
import org.code4everything.hutool.converter.DateConverter;
import org.code4everything.hutool.converter.FileConverter;
import org.code4everything.hutool.converter.ListStringConverter;
import org.code4everything.hutool.converter.PatternConverter;

import java.io.File;
import java.lang.reflect.Field;
import java.lang.reflect.Modifier;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.StringJoiner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

/**
 * @author pantao
 * @since 2020/10/30
 */
public final class Utils {

    private static JSONObject classAliasJson = null;

    private static JarClassLoader classLoader = null;

    private static List<String> mvnRepositoryHome = null;

    private Utils() {}

    public static String listFiles(@IOConverter(FileConverter.class) File file) {
        if (!FileUtil.exist(file)) {
            return "file not found!";
        }
        if (FileUtil.isFile(file)) {
            return DateUtil.formatDateTime(new Date(file.lastModified())) + "\t" + FileUtil.readableFileSize(file) + "\t" + file.getName();
        }

        List<String> filter = MethodArg.getSubParams(Hutool.ARG, 1);
        File[] files = file.listFiles((dir, name) -> {
            if (Utils.isCollectionEmpty(filter)) {
                return true;
            }
            String filename = name.toLowerCase();
            return filter.stream().anyMatch(filename::contains);
        });
        if (Objects.isNull(files) || files.length == 0) {
            return "";
        }

        Arrays.sort(files, ComparatorChain.of(Comparator.comparingInt(f -> f.isDirectory() ? 0 : 1), Comparator.comparing(File::getName), Comparator.comparingLong(File::lastModified)));
        StringJoiner joiner = new StringJoiner("\n");
        int maxLen = 0;
        String[] size = new String[files.length];
        for (int i = 0; i < files.length; i++) {
            size[i] = addSuffixIfNot(FileUtil.readableFileSize(files[i]).replace(" ", ""), "B");
            if (!Character.isDigit(size[i].charAt(size[i].length() - 2))) {
                size[i] = size[i].substring(0, size[i].length() - 1);
            }
            if (size[i].length() > maxLen) {
                maxLen = size[i].length();
            }
        }

        for (int i = 0; i < files.length; i++) {
            file = files[i];
            joiner.add(DateUtil.formatDateTime(new Date(file.lastModified())) + "  " + StrUtil.padPre(size[i], maxLen, " ") + "  " + file.getName());
        }

        return joiner.toString();
    }

    public static String dayProcess(@IOConverter(DateConverter.class) DateTime specificDate) {
        DateTime date = DateUtil.beginOfDay(specificDate);
        double todayProcess = (specificDate.getTime() - date.getTime()) * 100 / 24D / 60 / 60 / 1000;
        Week weekEnum = DateUtil.dayOfWeekEnum(date);
        int week = weekEnum.getValue() - 1;
        week = (week == 0 ? 7 : week) * 24 - 24;
        double weekProcess = (week + specificDate.hour(true)) * 100 / 7D / 24;
        double monthProcess = DateUtil.dayOfMonth(date) * 100 / (double) DateUtil.endOfMonth(date).dayOfMonth();
        double yearProcess = DateUtil.dayOfYear(date) * 100 / (double) DateUtil.endOfYear(date).dayOfYear();

        String template = String.format("%s %s %s%n", lunar(specificDate), weekEnum.toChinese("周"), Hutool.getSimpleDateFormat().format(specificDate));
        template += String.format("%n今日 [%s]: %05.2f%%", getDayProcessString(todayProcess), todayProcess);
        template += String.format("%n本周 [%s]: %05.2f%%", getDayProcessString(weekProcess), weekProcess);
        template += String.format("%n本月 [%s]: %05.2f%%", getDayProcessString(monthProcess), monthProcess);
        template += String.format("%n本年 [%s]: %05.2f%%", getDayProcessString(yearProcess), yearProcess);
        return template;
    }

    private static String getDayProcessString(@IOConverter double process) {
        int p = (int) (Math.ceil(process * 100) / 100);
        char[] cs = new char[100];
        Arrays.fill(cs, 0, p, 'o');
        Arrays.fill(cs, p, 100, ' ');
        return new String(cs);
    }

    public static String toHttpUrlString(Map<String, ?> paramMap) {
        if (paramMap == null || paramMap.isEmpty()) {
            return "";
        }

        StringBuilder sb = new StringBuilder();
        String sep = "?";
        for (Map.Entry<String, ?> entry : paramMap.entrySet()) {
            if (Objects.isNull(entry.getValue())) {
                continue;
            }
            String value = entry.getValue().toString();
            if (isStringEmpty(value)) {
                continue;
            }
            sb.append(sep).append(entry.getKey()).append("=").append(value);
            sep = "&";
        }
        return sb.toString();
    }

    public static String getFieldNames(@IOConverter(ClassConverter.class) Class<?> clazz) throws Exception {
        if (clazz.isPrimitive()) {
            clazz = parseClass0(parseClassName0("j." + clazz.getName()));
        }

        Field[] fields = ReflectUtil.getFields(clazz);
        StringJoiner joiner = new StringJoiner("\n");
        Map<String, List<String>> modifierMap = new HashMap<>();
        ComparatorChain<String> comparators = new ComparatorChain<>();
        comparators.addComparator(Comparator.comparingInt(o -> modifierMap.computeIfAbsent(o, s -> {
            String[] ss = s.split(" ");
            return ss.length > 2 ? Arrays.stream(ss).limit(ss.length - 2L).collect(Collectors.toList()) : Collections.emptyList();
        }).size()));
        comparators.addComparator(Comparator.naturalOrder());

        List<String> modifierList = new ArrayList<>();
        Holder<String> holder = Holder.of("");
        List<String> filter = MethodArg.getSubParams(Hutool.ARG, 1);

        Arrays.stream(fields).filter(e -> {
            if (Utils.isCollectionEmpty(filter)) {
                return true;
            }
            String fieldName = e.getName().toLowerCase();
            return filter.stream().anyMatch(fieldName::contains);
        }).map(field -> {
            String line = "";
            int modifiers = field.getModifiers();
            if (Modifier.isPrivate(modifiers)) {
                line += "private ";
            } else if (Modifier.isProtected(modifiers)) {
                line += "protected ";
            } else if (Modifier.isPublic(modifiers)) {
                line += "public ";
            }

            if (Modifier.isStatic(modifiers)) {
                line += "static ";
            }
            if (Modifier.isFinal(modifiers)) {
                line += "final ";
            }
            if (Modifier.isTransient(modifiers)) {
                line += "transient ";
            }
            if (Modifier.isVolatile(modifiers)) {
                line += "volatile ";
            }
            return line + field.getType().getSimpleName() + " " + field.getName();
        }).sorted(comparators).forEach(s -> {
            String line = s;
            List<String> list = modifierMap.get(s);
            if (!modifierList.equals(list)) {
                line = holder.get() + line;
                modifierList.clear();
                modifierList.addAll(list);
            }

            joiner.add(line);
            holder.set("\n");
        });

        return joiner.toString();
    }

    public static Object getStaticFieldValue(@IOConverter(ClassConverter.class) Class<?> clazz, String fieldName) throws Exception {
        if (clazz.isPrimitive()) {
            clazz = parseClass0(parseClassName0("j." + clazz.getName()));
        }
        Field field = ReflectUtil.getField(clazz, fieldName);
        return ReflectUtil.getStaticFieldValue(field);
    }

    @IOConverter(ListStringConverter.class)
    public static List<String> getMatchedItems(@IOConverter(PatternConverter.class) Pattern regex, String content) {
        ArrayList<String> result = new ArrayList<>();
        Matcher matcher = regex.matcher(content);
        while (matcher.find()) {
            result.add(matcher.group(0));
        }
        return result;
    }

    public static String getSupperClass(@IOConverter(ClassConverter.class) Class<?> clazz) {
        StringJoiner joiner = new StringJoiner("\n");
        getSupperClass(joiner, "", clazz);
        return joiner.toString();
    }

    private static void getSupperClass(StringJoiner joiner, String prefix, Class<?> clazz) {
        if (clazz == null || Object.class == clazz) {
            return;
        }

        joiner.add(prefix + (clazz.isInterface() ? "<>" : "") + clazz.getName());

        getSupperClass(joiner, prefix + "|    ", clazz.getSuperclass());
        for (Class<?> anInterface : clazz.getInterfaces()) {
            getSupperClass(joiner, prefix + "|    ", anInterface);
        }
    }

    public static boolean assignableFrom(@IOConverter(ClassConverter.class) Class<?> sourceClass, @IOConverter(ClassConverter.class) Class<?> testClass) {
        return sourceClass.isAssignableFrom(testClass);
    }

    public static String calc(String expression, @IOConverter int scale) {
        double res = Calculator.conversion(expression);
        return String.format("%." + scale + "f", res);
    }

    public static String lunar(@IOConverter(DateConverter.class) Date date) {
        return new ChineseDate(date).toString();
    }

    public static long date2Millis(@IOConverter(DateConverter.class) Date date) {
        return Objects.isNull(date) ? 0 : date.getTime();
    }

    public static String toUpperCase(String str) {
        return isStringEmpty(str) ? "" : str.toUpperCase();
    }

    public static String toLowerCase(String str) {
        return isStringEmpty(str) ? "" : str.toLowerCase();
    }

    public static String grep(@IOConverter(PatternConverter.class) Pattern pattern, @IOConverter(ListStringConverter.class) List<String> lines) {
        if (isCollectionEmpty(lines) && !isStringEmpty(Hutool.resultString)) {
            lines = new ListStringConverter().useLineSep().string2Object(Hutool.resultString);
        }
        StringJoiner joiner = new StringJoiner("\n");
        for (String line : lines) {
            if (pattern.matcher(line).find()) {
                joiner.add(line);
            }
        }
        return joiner.toString();
    }

    static <T> boolean isArrayEmpty(T[] arr) {
        return arr == null || arr.length == 0;
    }

    public static Class<?> parseClass(String className) throws Exception {
        switch (className) {
            case "bool":
            case "boolean":
                return boolean.class;
            case "byte":
                return byte.class;
            case "short":
                return short.class;
            case "char":
                return char.class;
            case "int":
                return int.class;
            case "long":
                return long.class;
            case "float":
                return float.class;
            case "double":
                return double.class;
            default:
                return parseClass0(className);
        }
    }

    private static Class<?> parseClass0(String className) throws Exception {
        className = parseClassName0(className);
        try {
            return Class.forName(className);
        } catch (Exception e) {
            // ignore
        }

        Hutool.debugOutput("loading class: " + className);
        if (Objects.isNull(classLoader)) {
            classLoader = new JarClassLoader();
            classLoader.addJar(FileUtil.file(Hutool.homeDir, "external"));
            File externalConf = FileUtil.file(Hutool.homeDir, "external.conf");
            if (FileUtil.exist(externalConf)) {
                String[] externalPaths = FileUtil.readUtf8String(externalConf).split(",");
                for (String externalPath : externalPaths) {
                    File external = parseClasspath(externalPath);
                    if (FileUtil.exist(external)) {
                        Hutool.debugOutput("add class path: " + external.getAbsolutePath());
                        classLoader.addURL(external);
                    }
                }
            }
        }

        return classLoader.loadClass(className);
    }

    private static File parseClasspath(String path) {
        path = path.trim();
        if (isStringEmpty(path)) {
            return null;
        }

        if (path.startsWith("mvn:")) {
            path = path.substring(4);
            if (isStringEmpty(path)) {
                return null;
            }
            String[] coordinates = path.split(":", 3);
            if (coordinates.length != 3) {
                Hutool.debugOutput("mvn coordinate format error: " + path);
                return null;
            }
            if (Objects.isNull(mvnRepositoryHome)) {
                mvnRepositoryHome = Arrays.asList("~", ".m2", "repository");
            }
            List<String> paths = new ArrayList<>(mvnRepositoryHome);
            paths.addAll(Arrays.asList(coordinates[0].split("\\.")));
            String name = coordinates[1];
            String version = coordinates[2];
            paths.add(name);
            paths.add(version);
            paths.add(name + "-" + version + ".jar");
            File file = FileUtil.file(paths.toArray(new String[0]));
            Hutool.debugOutput("get mvn path: " + file.getAbsolutePath());
            return file;
        }

        return FileUtil.file(path);
    }

    public static String parseClassName(String className) {
        switch (className) {
            case "bool":
            case "boolean":
                return "boolean";
            case "byte":
                return "byte";
            case "short":
                return "short";
            case "char":
                return "char";
            case "int":
                return "int";
            case "long":
                return "long";
            case "float":
                return "float";
            case "double":
                return "double";
            default:
                return parseClassName0(className);
        }
    }

    private static String parseClassName0(String className) {
        switch (className) {
            case "string":
                return "java.lang.String";
            case "j.char.seq":
                return "java.lang.CharSequence";
            case "file":
                return "java.io.File";
            case "charset":
                return "java.nio.charset.Charset";
            case "date":
                return "java.util.Date";
            case "class":
                return "java.lang.Class";
            case "j.bool":
            case "j.boolean":
                return "java.lang.Boolean";
            case "j.byte":
                return "java.lang.Byte";
            case "j.short":
                return "java.lang.Short";
            case "j.int":
            case "j.integer":
                return "java.lang.Integer";
            case "j.char":
                return "java.lang.Character";
            case "j.long":
                return "java.lang.Long";
            case "j.float":
                return "java.lang.Float";
            case "j.double":
                return "java.lang.Double";
            case "reg.pattern":
                return "java.util.regex.Pattern";
            case "map":
                return "java.util.Map";
            case "list":
                return "java.util.List";
            case "set":
                return "java.util.Set";
            default:
                return parseClassName00(className);
        }
    }

    private static String parseClassName00(String className) {
        if (className.endsWith(";") && className.contains("[L")) {
            int idx = className.indexOf("L");
            return className.substring(0, idx + 1) + parseClassName(className.substring(idx + 1, className.length() - 1)) + ";";
        }
        if (className.length() > 16) {
            return className;
        }
        if (classAliasJson == null) {
            classAliasJson = Hutool.getAlias("", Hutool.homeDir, Hutool.CLASS_JSON);
            classAliasJson.putAll(Hutool.getAlias("", "", Hutool.CLASS_JSON));
        }
        JSONObject classJson = classAliasJson.getJSONObject(className);
        if (classJson != null) {
            String className0 = classJson.getString(Hutool.CLAZZ_KEY);
            if (!isStringEmpty(className0)) {
                className = className0;
            }
        }
        return className;
    }

    public static boolean isStringEmpty(String str) {
        return str == null || str.length() == 0;
    }

    public static <T> boolean isCollectionEmpty(Collection<T> collection) {
        return collection == null || collection.isEmpty();
    }

    static String addPrefixIfNot(String str, String prefix) {
        if (isStringEmpty(str) || isStringEmpty(prefix) || str.startsWith(prefix)) {
            return str;
        }
        return prefix + str;
    }

    static String addSuffixIfNot(String str, String suffix) {
        if (isStringEmpty(str) || isStringEmpty(suffix) || str.endsWith(suffix)) {
            return str;
        }
        return str + suffix;
    }

    public static String padAfter(String str, int len, char pad) {
        if (str == null) {
            str = "";
        }

        int diff = len - str.length();
        if (diff < 1) {
            return str;
        }

        char[] cs = new char[diff];
        for (int i = 0; i < diff; i++) {
            cs[i] = pad;
        }

        return str + new String(cs);
    }

    public static String outputPublicStaticMethods(String className) {
        return isStringEmpty(className) ? "" : outputPublicStaticMethods0(className);
    }

    private static String outputPublicStaticMethods0(String className) {
        ClassPool pool = ClassPool.getDefault();
        StringJoiner joiner = new StringJoiner("\n");
        List<String> filter = MethodArg.getSubParams(Hutool.ARG, 1);

        try {
            CtClass ctClass = pool.get(parseClassName(className));
            CtMethod[] methods = ctClass.getMethods();
            List<String> lineList = new ArrayList<>(methods.length);
            for (CtMethod method : methods) {
                int modifiers = method.getModifiers();
                if (!Modifier.isPublic(modifiers) || !Modifier.isStatic(modifiers)) {
                    continue;
                }

                if (!isCollectionEmpty(filter)) {
                    String methodName = method.getName().toLowerCase();
                    if (filter.stream().noneMatch(methodName::contains)) {
                        continue;
                    }
                }

                lineList.add(getMethodFullInfo(method, null));
            }
            lineList.stream().sorted(String::compareTo).forEach(joiner::add);
        } catch (Exception e) {
            Hutool.debugOutput("parse class static methods error: %s", ExceptionUtil.stacktraceToString(e, Integer.MAX_VALUE));
        }

        return joiner.toString();
    }

    static String getMethodFullInfo(CtMethod method, Map<String, String> defaultValueMap) throws NotFoundException {
        CtClass[] parameterTypes = method.getParameterTypes();
        if (Hutool.isDebug()) {
            StringJoiner joiner = new StringJoiner(",");
            Arrays.stream(parameterTypes).forEach(e -> joiner.add(e.getName()));
            Hutool.debugOutput("parse method full info: %s#%s(%s)", method.getDeclaringClass().getName(), method.getName(), joiner);
        }

        StringJoiner paramJoiner = new StringJoiner(", ");
        LocalVariableAttribute attribute = (LocalVariableAttribute) method.getMethodInfo().getCodeAttribute().getAttribute(LocalVariableAttribute.tag);

        for (int i = 0; i < parameterTypes.length; i++) {
            CtClass parameterType = parameterTypes[i];
            String paramType = parameterType.getName();

            String paramStr = attribute.variableName(i) + ":" + paramType;
            if (defaultValueMap != null) {
                String defaultValue = defaultValueMap.get(paramType + i);
                if (!isStringEmpty(defaultValue)) {
                    paramStr += "=" + defaultValue;
                }
            }
            paramJoiner.add(paramStr);
        }

        return method.getName() + "(" + paramJoiner + ")";
    }
}
